### DISCUSSION TITLE
Lesson 13: gas cost

### DISCUSSION THREAD

**Discussion by Himchenhim** [Cyfrin/foundry-full-course-cu#1437] [General]
Why do we consider UUPS pattern more gas efficient than TransparentProxy pattern?

From openzeppelin docs:
"...While both of these share the same interface for upgrades, in UUPS proxies the upgrade is handled by the implementation, and can eventually be removed. Transparent proxies, on the other hand, include the upgrade and admin logic in the proxy itself. This means Transparent proxy is more expensive to deploy than what is possible with UUPS proxies."

I understand why it is less expensive deploy UUPS proxy (because we don't need to deploy management function in UUPS proxy).
However, we need to add `upgradeTo()` and `upgradeToAndCall()` after every upgrade  of implementation contract -> we just pay more gas for deploying Implementation contract + we will pay for `upgradeTo` and `upgradeToAndCall()` much more time than once using TransparentProxy pattern.

Where am I wrong? Could you please help me with that?

**Comment by EngrPips** [comment]
@Himchenhim  Thats indeed a good question to ask, here is what i found when i ask an AI buddy 

You are correct in noting that the UUPS pattern requires additional gas for deploying the implementation contract and for calling the upgradeTo and upgradeToAndCall functions after every upgrade. This does indeed incur more gas costs compared to the TransparentProxy pattern, which includes the upgrade and admin logic in the proxy itself.

The reason the UUPS pattern is considered more gas efficient than the TransparentProxy pattern lies in the long-term use of the proxies. While the initial deployment of a TransparentProxy may be cheaper, over time, the UUPS pattern becomes more efficient. This is because the UUPS pattern allows for the removal of upgrade and admin logic from the proxy after the initial deployment, thereby reducing the gas cost for subsequent transactions.

In contrast, the TransparentProxy pattern retains the upgrade and admin logic in the proxy itself, leading to higher gas costs for all subsequent transactions.

Therefore, the comparison of gas efficiency between these patterns takes into account the long-term usage and the reduction in gas costs for UUPS proxies after the initial deployment.
